src/utreexo.cairo

```diff
@@ -0,0 +1,111 @@
+    /// Index is the root height, None means a gap.
+    pub roots: Span<Option<felt252>>,
+    /// Total number of leaves (in the bottom-most row).
+    /// Required to calculate the number of nodes in a particular row.
+    /// Can be reconstructed from the roots, but cached for convenience.
+    pub num_leaves: u64,
+}
+
+/// Accumulator interface
+pub trait UtreexoAccumulator {
+    /// Adds single output to the accumulator.
+    /// The order *is important*: adding A,B and B,A would result in different states.
+    ///
+    /// Note that this call also pushes old UTXOs "to the left", to a larger subtree.
+    /// This mechanism ensures that short-lived outputs have small inclusion proofs.
+    fn add(ref self: UtreexoState, output: UtreexoOutput);
+
+    /// Verifies inclusion proof for a single output.
+    fn verify(
+        self: @UtreexoState, output: @UtreexoOutput, proof: @UtreexoProof
+    ) -> Result<(), UtreexoError>;
+
+    /// Removes single output from the accumlator (order is important).
+    ///
+    /// Note that once verified, the output itself is not required for deletion,
+    /// the leaf index plus inclusion proof is enough.
+    fn delete(ref self: UtreexoState, proof: @UtreexoProof);
+
+    /// Verifies batch proof for multiple outputs (e.g. all outputs in a block).
+    fn verify_batch(
+        self: @UtreexoState, outputs: Span<UtreexoOutput>, proof: @UtreexoBatchProof
+    ) -> Result<(), UtreexoError>;
+
+    /// Removes multiple outputs from the accumulator.
+    fn delete_batch(ref self: UtreexoState, proof: @UtreexoBatchProof);
+}
+
+#[derive(Drop, Copy, PartialEq)]
+pub enum UtreexoError {}
+
+/// TXO extended with info about parent transaction and the position within it.
+/// The hash of this structure is a leaf node in the Utreexo Merkle tree forest.
+#[derive(Drop, Copy)]
+pub struct UtreexoOutput {
+    /// The TXID this output belongs to.
+    pub txid: u256,
+    /// The index of this output.
+    pub vout: u32,
+    /// The height of the block this output was created at.
+    pub block_height: u32,
+    /// Flag indicating if this is a coinbase output.
+    pub is_coinbase: bool,
+    /// Transaction output.
+    pub output: TxOut,
+}
+
+/// Utreexo inclusion proof for a single transaction output.
+#[derive(Drop, Copy)]
+pub struct UtreexoProof {
+    /// Index of the leaf in the forest, but also an encoded binary path,
+    /// specifying which sibling node is left and which is right.
+    pub leaf_index: u64,
+    /// List of sibling nodes required to calculate the root.
+    pub proof: Span<felt252>,
+}
+
+/// Utreexo inclusion proof for multiple outputs.
+/// Compatible with https://github.com/utreexo/utreexo
+#[derive(Drop, Copy)]
+pub struct UtreexoBatchProof {
+    /// Indices of leaves to be deleted (ordered starting from 0, left to right).
+    pub targets: Span<u64>,
+    /// List of sibling nodes required to calculate the root.
+    pub proof: Span<felt252>,
```

Comment:
Shouldn't this be a tree like struct?

Reply:
The verification/deletion algorithms go row by row, so this is basically a linearization of the subset of forest nodes.